src/compiler.rs

```diff
@@ -34,26 +34,56 @@ impl<'ctx> Compiler<'ctx> {
         Target::initialize_all(&InitializationConfig::default());
     }
 
-    pub fn compile(&self, program: &[u8]) -> Result<(), String> {
-        let mut while_blocks = VecDeque::new();
+    pub fn compile(&self, program: String) -> Result<(), String> {
+        let mut while_blocks = VecDeque::with_capacity(1024);
         let functions = self.init_functions();
         let (data, ptr) = self.build_main(&functions);
         self.init_pointers(&functions, &data, &ptr)?;
 
-        let mut pc = 0;
-        while pc < program.len() {
-            match program[pc] as char {
-                '>' => self.build_add_ptr(1, &ptr),
-                '<' => self.build_add_ptr(-1, &ptr),
-                '+' => self.build_add(1, &ptr),
-                '-' => self.build_add(-1, &ptr),
+        let mut iter = program.chars().peekable();
+        let mut consecutive = 1;
+
+        while let Some(command) = iter.next() {
+            match command {
```

Comment:
I think LLVM is smart enough to optimize this for us, but I'll have to check. Either way, nice catch!